File: agent-factory/src/router.js
// Model routing + cost ledger. Picks the cheapest CAPABLE model per task:
// cheap/free tier for triage & simple work, stronger tiers only when the task's
// estimated complexity crosses the configured thresholds. Tracks running cost.
import { loadModels, costLog } from './util.js';

// Heuristic complexity estimate in 0..1 from task type + payload size/flags.
export function estimateComplexity(task) {
  const base = {
    prospect: 0.35,
    analyze: 0.55,
    propose: 0.7,
    triage: 0.1,
    classify: 0.15,
    summarize: 0.3,
    'build-frontend': 0.85,
    'build-cms': 0.9,
    'build-tutor': 0.65,
    'build-3d': 0.8,
    assemble: 0.4,
    report: 0.25,
  }[task.type] ?? 0.4;

  let score = base;
  const p = task.payload || {};
  if (p.high_stakes) score += 0.15;
  if (p.complexity) score = p.complexity; // explicit override
  const size = JSON.stringify(p).length;
  score += Math.min(0.15, size / 8000); // bigger payloads -> a bit harder
  if (task.attempts > 1) score += 0.1 * (task.attempts - 1); // escalate on retries
  return Math.max(0, Math.min(1, score));
}

export function pickModel(task, cfg) {
  const models = loadModels();
  const c = estimateComplexity(task);
  let tier = 1;
  if (c >= cfg.routing.escalate_to_tier3_at) tier = 3;
  else if (c >= cfg.routing.escalate_to_tier2_at) tier = 2;

  // candidates at the chosen tier
  let candidates = models.filter(m => m.tier === tier);
  // tier-1: optionally prefer the free stub model to maximize throughput-per-dollar
  if (tier === 1 && cfg.routing.prefer_free_stub) {
    const free = candidates.filter(m => m.input_per_m === 0 && m.output_per_m === 0);
    if (free.length) candidates = free;
  }
  if (!candidates.length) candidates = models.filter(m => m.tier <= tier);
  // cheapest by blended price
  candidates.sort((a, b) =>
    (a.input_per_m + a.output_per_m) - (b.input_per_m + b.output_per_m));
  const chosen = candidates[0];
  return { model: chosen, tier, complexity: Number(c.toFixed(3)) };
}

export function priceOf(modelId) {
  return loadModels().find(m => m.id === modelId);
}

// Compute USD cost from token usage and log it to COSTS.log (modeled, not charged).
export function charge({ task, model, usage, latency_ms, mode }) {
  const m = priceOf(model.id) || model;
  const cost =
    (usage.input_tokens / 1e6) * m.input_per_m +
    (usage.output_tokens / 1e6) * m.output_per_m;
  costLog({
    task_id: task.id, type: task.type, model: model.id, tier: model.tier,
    input_tokens: usage.input_tokens, output_tokens: usage.output_tokens,
    cost_usd: Number(cost.toFixed(6)), latency_ms, mode,
  });
  return Number(cost.toFixed(6));
}

export default { estimateComplexity, pickModel, charge, priceOf };

File: agent-factory/factory/router.py
"""Model routing + cost control.

Picks the cheapest *capable* model per task and tracks a running cost ledger in
COSTS.log. Optimizes throughput-per-dollar: cheap/local tiers handle triage and
simple work; the strong tier is only used when a task's complexity demands it.

Runs OFFLINE by default. A model "call" returns deterministic synthetic output
and an *estimated* cost. A real OpenRouter/local call is attempted only if BOTH
an API key is present AND FACTORY_ALLOW_LIVE_CALLS=1 — and even then it falls
back to stub on any error, so the whole system always runs end-to-end.
"""
from __future__ import annotations

import hashlib
import json
import time
import urllib.error
import urllib.request
from dataclasses import dataclass, asdict

from .paths import CONFIG_PATH, COSTS_LOG, load_env


@dataclass
class RouteResult:
    tier: str
    model: str
    provider: str
    mode: str            # "stub" | "live"
    input_tokens: int
    output_tokens: int
    cost_usd: float
    latency_ms: int
    text: str


def load_config() -> dict:
    return json.loads(CONFIG_PATH.read_text())


def _est_tokens(text: str) -> int:
    # ~4 chars/token heuristic, floor of 1.
    return max(1, len(text) // 4)


def required_capability(complexity: float) -> float:
    """Map a task's [0,1] complexity to the minimum model capability needed."""
    return max(0.0, min(1.0, complexity))


def choose_tier(complexity: float, cfg: dict, env: dict) -> tuple[str, dict]:
    """Cheapest tier whose capability covers the task. Respects cheap_first and
    the escalation threshold from tunables."""
    routing = cfg["routing"]
    tun = cfg["tunables"]
    need = required_capability(complexity)

    order = routing["tier_order"]
    models = routing["models"]

    # If complexity exceeds the escalation threshold, never start below "cheap":
    # local-triage is fine for sorting, not for the hard task itself.
    escalate = complexity >= tun["complexity_escalation_threshold"]

    candidates = order if tun.get("cheap_first", True) else list(reversed(order))
    for name in candidates:
        spec = models[name]
        if escalate and spec["max_capability"] < required_capability(0.5):
            continue
        if spec["max_capability"] >= need:
            return name, spec
    # Nothing covers it — use the strongest available.
    strongest = max(models.items(), key=lambda kv: kv[1]["max_capability"])
    return strongest[0], strongest[1]


def _stub_response(prompt: str, spec: dict, out_tokens: int) -> str:
    """Deterministic synthetic completion so offline runs are reproducible."""
    h = hashlib.sha256(prompt.encode()).hexdigest()[:8]
    return (f"[stub:{spec['model']}#{h}] processed {out_tokens}-token response "
            f"for prompt of {_est_tokens(prompt)} tokens")


def _post_json(url: str, body: bytes, headers: dict, timeout: float, retries: int):
    """POST and parse an OpenAI-compatible chat-completion response.

    Returns (text, usage_dict) or None on any failure. usage_dict carries real
    prompt/completion token counts when the endpoint reports them, so cost is
    measured rather than estimated.
    """
    last_err = None
    for attempt in range(retries + 1):
        try:
            req = urllib.request.Request(url, data=body, headers=headers)
            with urllib.request.urlopen(req, timeout=timeout) as resp:
                data = json.loads(resp.read())
            text = data["choices"][0]["message"]["content"]
            usage = data.get("usage", {}) or {}
            return text, usage
        except Exception as e:  # transient network / parse error -> retry then give up
            last_err = e
            if attempt < retries:
                time.sleep(0.25 * (attempt + 1))
    return None


def _live_openrouter(prompt: str, spec: dict, env: dict, timeout: float, retries: int):
    key = env.get("OPENROUTER_API_KEY", "")
    if not key:
        return None
    body = json.dumps({
        "model": spec["model"],
        "messages": [{"role": "user", "content": prompt}],
    }).encode()
    headers = {"Authorization": f"Bearer {key}", "Content-Type": "application/json"}
    base = env.get("OPENROUTER_BASE_URL", "https://openrouter.ai/api/v1")
    return _post_json(base.rstrip("/") + "/chat/completions",
                      body, headers, timeout, retries)


def _live_local(prompt: str, spec: dict, env: dict, timeout: float, retries: int):
    base = env.get("LOCAL_MODEL_URL", "")
    if not base:
        return None
    body = json.dumps({
        "model": env.get("LOCAL_MODEL_NAME", spec["model"]),
        "messages": [{"role": "user", "content": prompt}],
    }).encode()
    return _post_json(base.rstrip("/") + "/chat/completions",
                      body, {"Content-Type": "application/json"}, timeout, retries)


def route_and_run(prompt: str, complexity: float,
                  expected_output_tokens: int = 220) -> RouteResult:
    cfg = load_config()
    env = load_env()
    tier, spec = choose_tier(complexity, cfg, env)

    in_tok = _est_tokens(prompt)
    out_tok = expected_output_tokens

    live_allowed = env.get("FACTORY_ALLOW_LIVE_CALLS") == "1"
    timeout = float(cfg["tunables"].get("live_call_timeout_seconds", 20))
    retries = int(cfg["tunables"].get("live_call_retries", 1))
    mode = "stub"
    text = None
    t0 = time.time()

    if live_allowed:
        out = None
        if spec["provider"] == "openrouter":
            out = _live_openrouter(prompt, spec, env, timeout, retries)
        elif spec["provider"] == "local":
            out = _live_local(prompt, spec, env, timeout, retries)
        if out is not None:
            text, usage = out
            mode = "live"
            # Prefer the endpoint's reported token usage; fall back to estimate.
            in_tok = int(usage.get("prompt_tokens") or in_tok)
            out_tok = int(usage.get("completion_tokens") or _est_tokens(text))

    if text is None:
        text = _stub_response(prompt, spec, out_tok)
        # tiny deterministic delay so latency metrics are non-zero
        time.sleep(0.01 + (in_tok % 7) * 0.002)

    latency_ms = int((time.time() - t0) * 1000)
    cost = (in_tok / 1000.0) * spec["cost_per_1k_input"] + \
           (out_tok / 1000.0) * spec["cost_per_1k_output"]

    result = RouteResult(
        tier=tier, model=spec["model"], provider=spec["provider"], mode=mode,
        input_tokens=in_tok, output_tokens=out_tok, cost_usd=round(cost, 6),
        latency_ms=latency_ms, text=text,
    )
    _log_cost(result)
    return result


def _log_cost(r: RouteResult) -> None:
    entry = {"ts": round(time.time(), 3), **asdict(r)}
    entry.pop("text", None)  # keep ledger compact
    COSTS_LOG.parent.mkdir(parents=True, exist_ok=True)
    with COSTS_LOG.open("a") as f:
        f.write(json.dumps(entry) + "\n")
